README.md
[![Gem Version](https://badge.fury.io/rb/gull.svg)](http://badge.fury.io/rb/gull)
[![Build Status](https://travis-ci.org/sethdeckard/gull.svg?branch=master)](https://travis-ci.org/sethdeckard/gull)
[![Coverage Status](https://coveralls.io/repos/sethdeckard/gull/badge.svg?branch=master)](https://coveralls.io/r/sethdeckard/gull)
[![Code Climate](https://codeclimate.com/github/sethdeckard/gull/badges/gpa.svg)](https://codeclimate.com/github/sethdeckard/gull)
[![Dependency Status](https://gemnasium.com/sethdeckard/gull.svg)](https://gemnasium.com/sethdeckard/gull)
[![security](https://hakiri.io/github/sethdeckard/gull/master.svg)](https://hakiri.io/github/sethdeckard/gull/master)
# Gull

Ruby client for parsing NOAA/NWS alerts, warnings, and watches. The name comes from the type of bird featured on the NOAA logo. Please read the Notes/Caveats section for limitations.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'gull'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install gull

## Usage

```ruby
require 'gull'

alerts = Gull::Alert.fetch
alert = alerts.first

alert.id
alert.alert_type
alert.title
alert.summary
alert.effective_at
alert.expires_at
alert.published_at
alert.area
alert.polygon
alert.geocode.fips6
alert.geocode.ugc
alert.urgency
alert.severity
alert.certainty
alert.vtec
```

To get alerts for a single state, territory, or marine zone just pass an optional URL

```ruby
oklahoma_url = 'http://alerts.weather.gov/cap/ok.php?x=1'
alerts = Gull::Alert.fetch(url: oklahoma_url)
```

You can also generate a map (a really long URL pointing to a map) of the polygon if alert has one (requires Google Static Maps API Key)

```ruby
alert.polygon.image_url 'YOUR_GOOGLE_API_KEY'

=> "http://maps.googleapis.com/maps/api/staticmap?size=640x640&maptype=roadmap&path=color:0xff0000|weight:3|fillcolor:0xff000060|38.73,-94.22|38.75,-94.16|38.57,-93.94|38.4,-93.84|38.4,-93.91|38.73,-94.22&key=YOUR_GOOGLE_API_KEY"
```

Options can be passed for map to override defaults

```ruby
options = { width: 600, height: 300, color: '0xfbf000', weight: 4,
            fillcolor: '0xfbf00070', maptype: 'hybrid' }
alert.polygon.image_url 'YOUR_GOOGLE_API_KEY', options
```

##Notes, Caveats
This library provides a simplified/flattened model of the [Common Alerting Protocol](http://docs.oasis-open.org/emergency/cap/v1.2/CAP-v1.2-os.html) based only on the elements NWS utilizes in their [public Atom feeds](http://alerts.weather.gov/). If you need a complete CAP parser I suggest looking at [RCAP](https://github.com/farrel/RCAP).

The NWS will often cancel or update alerts before their expiration time. The public Atom feeds only provide current active alerts and do not include these separate update and cancellation CAP messages.

The public Atom feeds have not always been reliable in terms of uptime and are often delayed by 2-3 minutes. If you are considering using this (or another gem/library that accesses the public Atom feeds) for mission critical purposes then you should explore other options.

### Urgency

| Symbol        | Definition          
| :------------- |:-------------
| :immediate  | Responsive action should betaken immediately
| :expected  | Responsive action should be taken soon (within next hour)
| :future  | Responsive action should be taken in the near future
| :past  | Responsive action is no longer required
| :unknown  | Urgency not known

### Severity

| Symbol        | Definition          
| :------------- |:-------------
| :extreme  | Extraordinary threat to life or property
| :severe  | Significant threat to life or property
| :moderate  | Possible threat to life or property
| :minor  | Minimal threat to life or property
| :unknown  | Severity unknown

### Certainty

| Symbol        | Definition          
| :------------- |:-------------
| :very_likely  | Highly likely (p > ~ 85%) or certain
| :likely  | Likely (p > ~50%)
| :possible  | Possible but not likely (p <= ~50%)
| :unlikely  | Not expected to occur (p ~ 0)
| :unknown  | Certainty unknown


## Contributing

1. Fork it ( https://github.com/sethdeckard/gull/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Run the specs, make sure they pass and that new features are covered. Code coverage should be 100%.
4. Commit your changes (`git commit -am 'Add some feature'`)
5. Push to the branch (`git push origin my-new-feature`)
6. Create a new Pull Request

## License

Gull is released under the [MIT License](http://www.opensource.org/licenses/MIT).


CHANGELOG.md
0.4.0 (07/26/2016) - Merged pull request #2 from [schrockwell](https://github.com/schrockwell), which adds `to_wkt` to `Polygon` which formats the polygon points as Well-Known Text (WKT). Removed `centroid` method, use new `to_wkt` method with an external geospatial library instead.

***

0.3.4 (12/16/2015) - Fixed parser to handle empty geocode.

***

0.3.3 (11/25/2015) - Handle additional service failure scenarios.

***

0.3.2 (11/10/2015) - Added BadResponseError to client error handling.

***

0.3.1 (6/25/2015) - Added nested error handling for other HTTPClient errors. HttpError has original attribute exposing the root error.

***

0.3.0 (6/23/2015) - Added Client class for error handling (bad entries accessed via errors array). Exposed option to turn on strict XML parsing for debugging. Alert.fetch uses new Client but is backwards compatible, however it doesn't give you access to errors array.

***

0.2.13 (6/20/2015) - Handle missing cap section

***

0.2.12 (6/12/2015) - Fix issue with missing link elements

***

0.2.11 (5/29/2015) -Added method to get original polygon string

***

0.2.10 (3/23/2015) - Added option to override default alert service URI.

***

0.2.1 (10/02/2014) - Added static map image to polygons.

***

0.2.0 (10/02/2014) - Introduced Polygon type.
 
***

0.1.1 (10/01/2014) - Refactored and simplified alert processing.

***

0.1.0 (10/01/2014) - Initial release, basic functionality of fetching and parsing alerts.


spec/alert_spec.rb
require 'spec_helper'

describe Gull::Alert do
  it 'should initialize with geocode' do
    alert = Gull::Alert.new
    expect(alert.geocode).not_to be_nil
  end

  it 'should fetch parsed alerts' do
    xml = File.read 'spec/fixtures/alerts.xml'

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    alerts = Gull::Alert.fetch
    expect(alerts.size).to eq(3)

    first = alerts.first
    expect(first.id).to eq 'http://alerts.weather.gov/cap/wwacapget.php?x=CA125171381DD0.HeatAdvisory'
    expect(first.link).to eq 'http://alerts.weather.gov/cap/wwacapget.php?x=CA125171381DD0.HeatAdvisory'
    expect(first.alert_type).to eq 'Heat Advisory'
    expect(first.title).to eq 'Heat Advisory issued October 01 at 8:40AM PDT' \
      ' until October 03 at 9:00PM PDT by NWS'
    expect(first.summary).to eq 'SUMMARY TEXT'

    coordinates = [[27.35, -81.79], [27.14, -81.89], [27.04, -81.97],
                   [27.04, -82.02], [27.14, -81.97], [27.35, -81.86],
                   [27.35, -81.79]]
    expect(first.polygon.coordinates).to eq coordinates

    expect(first.effective_at).to eq Time.parse('2014-10-01T08:40:00-07:00')
    expect(first.expires_at).to eq Time.parse('2014-10-03T21:00:00-07:00')
    expect(first.updated_at).to eq Time.parse('2014-10-01T08:40:00-07:05')
    expect(first.published_at).to eq Time.parse('2014-10-01T08:40:00-07:06')

    expect(first.area).to eq 'Southern Salinas Valley, Arroyo Seco and Lake ' \
      'San Antonio'
    expect(first.urgency).to eq :expected
    expect(first.severity).to eq :minor
    expect(first.certainty).to eq :very_likely

    expect(first.geocode.fips6).to be_nil
    expect(first.geocode.ugc).to be_nil

    expect(first.vtec).to eq '/O.NEW.KMTR.HT.Y.0002.141002T1900Z-141004T0400Z/'

    second = alerts[1]
    expect(second.polygon).to be_nil

    expect(second.geocode.fips6).to eq '006001 006013 006041 006053 006055 ' \
      '006069 006075 006081 006085 006087 006097'
    expect(second.geocode.ugc).to eq 'CAZ006 CAZ505 CAZ506 CAZ507 CAZ508 ' \
      'CAZ509 CAZ510 CAZ511 CAZ512 CAZ513 CAZ516 CAZ517 CAZ518 CAZ528 ' \
      'CAZ529 CAZ530'

    expect(second.vtec).to be_nil

    third = alerts[2]
    expect(third.vtec).to be_nil
    expect(third.link).to be_nil
  end

  it 'should fetch from url in options' do
    xml = File.read 'spec/fixtures/alerts.xml'

    stub_request(:get, 'http://test.url')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    alerts = Gull::Alert.fetch(url: 'http://test.url')
    expect(alerts.size).to eq(3)
  end

  it 'should enable strict xml parsing via option' do
    xml = File.read 'spec/fixtures/bad.xml'

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    expect { Gull::Alert.fetch(strict: true) }
      .to raise_error(Nokogiri::XML::SyntaxError)
  end

  it 'should handle empty alerts' do
    xml = File.read 'spec/fixtures/empty.xml'

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    alerts = Gull::Alert.fetch
    expect(alerts.size).to eq(0)
  end

  it 'should handle bad response' do
    xml = File.read 'spec/fixtures/bad.xml'

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    alerts = Gull::Alert.fetch
    expect(alerts.size).to eq(0)
  end

  it 'should handle missing cap section' do
    xml = File.read 'spec/fixtures/missing_cap.xml'

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    alerts = Gull::Alert.fetch
    expect(alerts.size).to eq 0
  end
end


spec/error_spec.rb
require 'spec_helper'

describe Gull::HttpError do
  it 'should instantiate and set original with current exception' do
    error = StandardError.new 'inner'
    begin
      fail error
    rescue StandardError
      http_error = Gull::HttpError.new 'test'
      expect(http_error.original).to eq error
      expect(http_error.original.message). to eq 'inner'
    end
  end
end


spec/client_spec.rb
require 'spec_helper'

describe Gull::Client do
  it 'should initialize with options' do
    xml = File.read 'spec/fixtures/alerts.xml'
    stub_request(:get, 'http://test.url')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    options = { url: 'http://test.url', strict: true }
    client = Gull::Client.new(options)
    alerts = client.fetch
    expect(alerts.size).to eq 3

    xml = File.read 'spec/fixtures/bad.xml'
    stub_request(:get, 'http://test.url')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    expect { client.fetch }
      .to raise_error(Nokogiri::XML::SyntaxError)
  end

  it 'should fetch alerts without options' do
    xml = File.read 'spec/fixtures/alerts.xml'

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    client = Gull::Client.new
    alerts = client.fetch
    expect(alerts.size).to eq 3
    expect(client.errors.size).to eq 0
  end

  it 'should handle incomplete entries in xml' do
    xml = File.read 'spec/fixtures/missing_cap.xml'

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_return(status: 200, body: xml, headers: {})

    client = Gull::Client.new
    alerts = client.fetch
    expect(alerts.size).to eq 0
    expect(client.errors.size).to eq 1
  end

  it 'should raise own error if timeout occurs' do
    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_timeout

    message = 'Timeout while connecting to NWS web service'
    client = Gull::Client.new
    expect { client.fetch }.to raise_error(Gull::TimeoutError, message)
  end

  it 'should raise own error if http errors occur' do
    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_raise(HTTPClient::KeepAliveDisconnected)

    message = 'Could not connect to NWS web service'
    client = Gull::Client.new
    expect { client.fetch }.to raise_error(Gull::HttpError, message) do |error|
      expect(error.original).to be_a(HTTPClient::KeepAliveDisconnected)
    end

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_raise(HTTPClient::BadResponseError)

    message = 'Could not connect to NWS web service'
    client = Gull::Client.new
    expect { client.fetch }.to raise_error(Gull::HttpError, message) do |error|
      expect(error.original).to be_a(HTTPClient::BadResponseError)
    end
  end

  it 'should raise own error if connection errors occur' do
    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_raise(SocketError)

    message = 'Could not connect to NWS web service'
    client = Gull::Client.new
    expect { client.fetch }.to raise_error(Gull::HttpError, message) do |error|
      expect(error.original).to be_a(SocketError)
    end

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_raise(Errno::ECONNREFUSED)

    message = 'Could not connect to NWS web service'
    client = Gull::Client.new
    expect { client.fetch }.to raise_error(Gull::HttpError, message) do |error|
      expect(error.original).to be_a(Errno::ECONNREFUSED)
    end

    stub_request(:get, 'http://alerts.weather.gov/cap/us.php?x=1')
      .with(headers: { 'Accept' => '*/*' })
      .to_raise(Errno::ECONNRESET)

    message = 'Could not connect to NWS web service'
    client = Gull::Client.new
    expect { client.fetch }.to raise_error(Gull::HttpError, message) do |error|
      expect(error.original).to be_a(Errno::ECONNRESET)
    end
  end
end


lib/gull/alert.rb
require 'httpclient'
require 'nokogiri'

module Gull
  # Gull represents an NWS/NOAA alert and provides the ability to fetch
  # them from the public web service
  class Alert
    attr_accessor :id, :title, :summary, :link, :alert_type, :polygon, :area,
                  :effective_at, :expires_at, :updated_at, :published_at,
                  :urgency, :severity, :certainty, :geocode, :vtec

    def initialize
      self.geocode = Geocode.new
    end

    def self.fetch(options = {})
      client = Client.new options
      client.fetch
    end

    def parse(element)
      parse_core_attributes element
      parse_times element
      parse_categories element

      parse_polygon element.xpath('cap:polygon').inner_text
      parse_geocode element.xpath('cap:geocode')
      parse_vtec element.xpath('cap:parameter')
    end

    private

    def parse_core_attributes(element)
      self.id = element.css('id').inner_text
      self.title = element.css('title').inner_text
      self.summary = element.css('summary').inner_text
      self.link = parse_link element
      self.alert_type = element.xpath('cap:event').inner_text
      self.area = element.xpath('cap:areaDesc').inner_text
    end

    def parse_link(element)
      link = element.css('link').first
      link.attributes['href'].value unless link.nil?
    end

    def parse_times(element)
      self.updated_at = Time.parse(element.css('updated').inner_text)
      self.published_at = Time.parse(element.css('published').inner_text)
      self.effective_at = Time.parse(element.xpath('cap:effective').inner_text)
      self.expires_at = Time.parse(element.xpath('cap:expires').inner_text)
    end

    def parse_categories(element)
      self.urgency = code_to_symbol element.xpath('cap:urgency').inner_text
      self.severity = code_to_symbol element.xpath('cap:severity').inner_text
      self.certainty = code_to_symbol element.xpath('cap:certainty').inner_text
    end

    def parse_polygon(text)
      return if text.empty?
      self.polygon = Polygon.new text
    end

    def parse_geocode(element)
      return if element.children.css('value').first.nil?

      geocode.fips6 = element.children.css('value').first.inner_text
      geocode.ugc = element.children.css('value').last.inner_text
    end

    def parse_vtec(element)
      value = element.children.css('value').inner_text
      self.vtec = value.empty? ? nil : value
    end

    def code_to_symbol(code)
      code.tr(' ', '_').downcase.to_sym
    end
  end
end


lib/gull/client.rb
require 'httpclient'
require 'nokogiri'

module Gull
  # Client exposes methods and options for fetching alerts from the NWS/NOAA
  # web service
  class Client
    attr_accessor :errors

    def initialize(options = {})
      @options = {
        url: 'http://alerts.weather.gov/cap/us.php?x=1',
        strict: false
      }.merge options
    end

    def fetch
      self.errors = []
      content = response
      document = Nokogiri::XML content do |config|
        config.strict if @options[:strict]
      end
      process document.xpath('//xmlns:feed/xmlns:entry', namespaces)
    end

    private

    def response
      client = HTTPClient.new
      begin
        return client.get_content @options[:url]
      rescue HTTPClient::TimeoutError
        raise TimeoutError, 'Timeout while connecting to NWS web service'
      rescue HTTPClient::KeepAliveDisconnected, HTTPClient::BadResponseError,
             SocketError, Errno::ECONNREFUSED, Errno::ECONNRESET
        raise HttpError, 'Could not connect to NWS web service'
      end
    end

    def process(entries)
      alerts = []
      entries.each do |entry|
        alert = create_instance entry
        alerts.push alert unless alert.nil?
        errors.push entry if alert.nil?
      end

      alerts
    end

    def create_instance(entry)
      return if entry.xpath('cap:event').empty?

      alert = Alert.new
      alert.parse entry
      alert
    end

    def namespaces
      { 'xmlns' => 'http://www.w3.org/2005/Atom',
        'cap' => 'urn:oasis:names:tc:emergency:cap:1.1',
        'ha' => 'http://www.alerting.net/namespace/index_1.0' }
    end
  end
end


lib/gull/polygon.rb
module Gull
  class Polygon
    attr_accessor :coordinates

    def initialize(polygon)
      self.coordinates = polygon.split(' ')
                                .map { |point| point.split(',').map(&:to_f) }
    end

    def image_url(api_key, options = {})
      options = {
        width: 640,
        height: 640,
        color: '0xff0000',
        weight: 3,
        fillcolor: '0xff000060',
        maptype: 'roadmap'
      }.merge(options)

      url_base = 'http://maps.googleapis.com/maps/api/staticmap'
      "#{url_base}?size=#{options[:width]}x#{options[:height]}" \
      "&maptype=#{options[:maptype]}&path=color:#{options[:color]}" \
      "|weight:#{options[:weight]}|fillcolor:#{options[:fillcolor]}" \
      "|#{coordinates_piped}&key=#{api_key}"
    end

    def to_s
      coordinates.map { |pair| pair.join(',') }.join(' ')
    end

    # Returns well-known text (WKT) formatted polygon
    def to_wkt
      pairs = coordinates.map { |pair| "#{pair.last} #{pair.first}" }
                         .join(', ')
      "POLYGON((#{pairs}))"
    end

    private

    def coordinates_piped
      coordinates.map { |pair| pair.join ',' }.join '|'
    end
  end
end
